### PR TITLE
Fix the issue when babel presets were not found.

### PR DESCRIPTION
**Pull request by paulmillr**

### CODE REVIEW

**Comment by paulmillr** [comment]
Folks, could you give me push access here please? Also, NPM.

**Comment by paulmillr** [comment]
@es128

**Comment by chrismccord** [comment]
@paulmillr What will it take to get source maps back in place? Also thanks so for much for tackling this issue!

**Comment by paulmillr** [comment]
@chrismccord a babel option that would allow to specify where the `node_modules` with babel presets are.

I suggest to use my fork of `babel-brunch` until folks would merge the PR in.

**Comment by chrismccord** [comment]
> a babel option that would allow to specify where the node_modules with babel presets are

Is there an open issue where I can track this? I can't get a handle on their phabricator system

**Comment by paulmillr** [comment]
> Is there an open issue where I can track this

We'll need to open one, yep.

**Comment by es128** [comment]
> Folks, could you give me push access here please? Also, NPM.

done

**Comment by es128** [comment]
Have you considered setting the es2015 preset package as a peer dependency here? Might solve the problem for folks on npm 2.x.

I caution against proceeding with this PR, as it looks like it will just trade one set of users' problem with another (anyone actively using the sourcemaps).

I don't see why babel does not use `require.resolve()`, as that would appear to be the more proper solution, but in the meantime, I'd try out adding it as a peer dependency (possibly both peer and regular dependency simultaneously).

**Comment by paulmillr** [comment]
> it looks like it will just trade one set of users' problem with another

at least folks would be able to run it. Bad sourcemaps is the least of all evils

**Comment by paulmillr** [comment]
Again, the problem is not related to npm v2. It's npm v3 too. In some cases.

**Comment by es128** [comment]
Have you tried out the idea of solving this by adding the preset as a peer dependency?

**Comment by chrismccord** [comment]
A maybe related issue I am having and has been reported by others:

1. Use babel 2.1.1 with npm support enabled
2. `npm install --save babel-preset-react`
3. Receive unhelpful error
```console
$ brunch build
06 Jan 14:12:04 - error: Initialization error - Error: ENOENT: no such file or directory, open 'path'
  at Error (native)
```

In desperation, I solved this issue by moving `babel-preset-react` to devDependencies, which still allows it to work for my usecase in Phoenix. Is this issue related? Any better fix for folks?

Example project with workaround:
https://github.com/chrismccord/phoenix_brunch_react/blob/master/package.json

**Comment by paulmillr** [comment]
@es128 npm3 removed "mandatory" peer deps. Warnings are unhelpful for users.

>We will also be changing the behavior of peerDependencies in npm@3. We won’t be automatically downloading the peer dependency anymore. Instead, we’ll warn you if the peer dependency isn’t already installed. This requires you to resolve peerDependency conflicts yourself, manually, but in the long run this should make it less likely that you’ll end up in a tricky spot with your packages’ dependencies. - http://blog.npmjs.org/post/110924823920/npm-weekly-5

**Comment by es128** [comment]
I disagree that it will be unhelpful, although admittedly I haven't tried it to confirm my assumptions. On npm2, where the issue is much more prevalent, the peer dependency will be installed, issue solved. On npm3, having it as both a dependency and a peer dependency should mean that it gets installed flat and suppress the warning, which also solves the problem. If the warning does occur, it tells users exactly how to solve the problem.

Under what circumstances does the problem to occur with npm3 btw? Maybe I'm not fully understanding all the situations involved.

Anyway, if you feel this is pressing you do have access now to do as you see fit icymi.

**Comment by paulmillr** [comment]
@chrismccord we're on this.

**Comment by goshakkk** [comment]
@chrismccord as `babel-preset-react` isn't a front-end dependency and npm for front-end integration isn't yet clever enough to actually figure out which packages are used in the front-end (in the works!), the solution for that small application would be to either whitelist the front-end packages (`config.npm.whitelist = ['react', 'react-dom', ...]`), or blacklist non-front-end packages (`config.npm.blacklist = ['babel-preset-react', ...]`)

**Comment by paulmillr** [comment]
tl;dr Brunch 2.1 requires `config.npm.whitelist`, Brunch 2.2 would figure out the stuff automatically.

**Comment by josevalim** [comment]
@es128 here is the issue that happens on node@3:

https://github.com/phoenixframework/phoenix/issues/1410#issuecomment-167974086

The steps to reproduce can be found here:

https://github.com/phoenixframework/phoenix/issues/1410#issuecomment-167975267

So it is *not* a problem of flat or tree dependencies, it is an issue of babel/brunch just looking for the dependencies in the wrong place.

**Comment by chrismccord** [comment]
@goshakkk @paulmillr ahh this makes sense. Thanks for clarify!

**Comment by es128** [comment]
So actually it looks like we're looking at 2 entirely different problems at the same time. 

One part is about babel's failure to use `require.resolve` for its plugins, which is what this PR intends to address and for which I've suggested the alternate possibility of using `peerDependencies`. 

The other is about brunch's npm integration being somewhat manual/naive still, which @paulmillr and @goshakkk are working on improving.

**Comment by es128** [comment]
Another possibility to consider is that babel-brunch > 6.0 just stops trying to work with default settings out of the box, while users are given instructions to install and specify the plugins/presets they intend to use, much like using Babel itself. Older versions could still be used and "just work"

**Comment by paulmillr** [comment]
@es128 it would solve the issue; although it's a bit non-Brunchy to expect users to install tons of plugins to just make stuff work.

As a user, I would really love to just install `es6-brunch` without messing with configs and shit. Or, `autoprefixer-brunch` (sadly, the owner of the repo deprecated it).

Not sure how this could work when there are more than 1 babel plugins, but sth to think about.

**Comment by es128** [comment]
I'd say it's a choice Babel has made for their users. I agree it goes against what we try to do with Brunch generally. But we tried to make easy defaults, and it hasn't worked out. May be better at this point to just get out of the way and document clearly for users the steps they'll need to take.

**Comment by chrismccord** [comment]
Let's hold on this for a moment, as @josevalim may have found a nice workaround by making phoenix.js a local npm dep

**Comment by josevalim** [comment]
We have fixed this on Phoenix by making the extra directory dependencies (at ../../deps/phoenix) simply be npm dependencies. We have added them to package.json as file dependencies and did a couple changes and now everything works! The initial PR is here: https://github.com/phoenixframework/phoenix/pull/1456

It also becomes a nice opportunity to show how the npm integration works, which is a :+1:. This would not have been possible without the work of the Brunch team on npm integration, so thank you so much.

I think we can close this issue as it no longer blocks Phoenix and losing source maps would be undesired. :)
